### PR TITLE
Rethink ChainId

### PR DESCRIPTION
**Pull request by michalsidzej**
Resolves L2B-2175
Resolves L2B-2081

This PR is a part of the work that will allow us to move the discovery package to a separate repository.
The scope of this PR is decoupling chainId from `backend` and `discovery` specific config values. 

1. `minTimestamp` has been moved close to configs in `backend` package. This seems quite logical.
2. `dethDomain` has been removed, as this feature is not used that much (according to my research in the [research channel](https://discord.com/channels/988352096178946142/1077549098158264441/1143231991022633030).  
3. `DiscoveryEtherscanLikeClient` has been created. It's only purpose is to provide `UnixTime(0)` to the constructor of `EtherscanLikePattern`, so we do not need to provide `minTimestamp` to discovery package config (it is not used anywhere in discovery logic, as discovery only uses `getContractSource` and `getContractDeploymentTx` which are not dependant on `minTimestamp`. It is not a pretty pattern but works well.
4. Refactor `ChainId` branded type

### CODE REVIEW

File: packages/backend/src/config/chains.ts
Comment (antooni) [inline packages/backend/src/config/chains.ts:16]: getChainMinTimestamp

Comment (michalsidzej) [reply packages/backend/src/config/chains.ts:16]: I mean chain is the argument so it speaks for itself, no?

**Review by antooni — CHANGES_REQUESTED** [review]
please resolve comments

File: packages/discovery/src/discovery/output/toDiscoveryOutput.ts
Comment (sdlyy) [inline packages/discovery/src/discovery/output/toDiscoveryOutput.ts:105]: Maybe we should do an exhaustive switch check here to clearly inform developers something's ain't right with deth targets? Assuming each chain should have related deth available

Comment (sdlyy) [reply packages/discovery/src/discovery/output/toDiscoveryOutput.ts:105]: `assertUnreachable`

File: packages/discovery/src/discovery/output/toDiscoveryOutput.ts
Comment (michalsidzej) [reply packages/discovery/src/discovery/output/toDiscoveryOutput.ts:89]: I'll remove it, there are better ways of using our discovery.

File: packages/backend/src/config/chains.ts
Comment (antooni) [reply packages/backend/src/config/chains.ts:16]: At one hand yes, but on the other I can very easily see the naming collision in the future which will make stating imports harder. I leave it up to you, I am not really pressing this change

File: packages/discovery/src/discovery/output/toDiscoveryOutput.ts
Comment (michalsidzej) [reply packages/discovery/src/discovery/output/toDiscoveryOutput.ts:105]: This will not work with `ChainId` as it extends string

Comment (sdlyy) [reply packages/discovery/src/discovery/output/toDiscoveryOutput.ts:105]: oh right - unfortunate

File: packages/shared-pure/src/types/ChainId.ts
Comment (michalsidzej) [reply packages/shared-pure/src/types/ChainId.ts:12]: ChainId.fromName() - creates chainId (number) from chain name
ChainId.fromNameTypesafe() - creates chainId from name, but will not allow you to use non existing chain name. Useful when creating ChainId.ETHEREUM (and other) constants, ts will shout at you in the editor if you misspell.
ChainId.getName(chainId) - gives you the name of the chain with id that you pass. Useful ex. when we save discovered.json, or any logging/errors.
ChainId.getAll() - return all chains. Used for testing. This is something worth rethinking actually.

File: packages/discovery/src/discovery/output/toDiscoveryOutput.ts
Comment (michalsidzej) [reply packages/discovery/src/discovery/output/toDiscoveryOutput.ts]: 04217afd

File: packages/shared-pure/src/types/ChainId.ts
Comment (michalsidzej) [reply packages/shared-pure/src/types/ChainId.ts:14]: Ad. 2. f329ed8d

File: packages/discovery/src/discovery/provider/DiscoveryEtherscanLikeClient.ts
Comment (michalsidzej) [reply packages/discovery/src/discovery/provider/DiscoveryEtherscanLikeClient.ts:4]: > On the other hand, we can add a static construction method to wrap minimum timestamp and spare us yet another class. It will simply partially apply rest arguments with fixed timestamp

I like this one more.

File: packages/discovery/src/discovery/engine/DiscoveryEngine.ts
Comment (antooni) [inline packages/discovery/src/discovery/engine/DiscoveryEngine.ts:10]: should this comment be there, the logic version has not been changed

**Review by antooni — COMMENTED** [review]
left small comments

**Review by antooni — COMMENTED** [review]
left small comments

**Review by antooni — COMMENTED** [review]
left small comments

File: packages/discovery/src/discovery/engine/DiscoveryEngine.ts
Comment (michalsidzej) [reply packages/discovery/src/discovery/engine/DiscoveryEngine.ts:10]: My bad, the version should be changed

Comment (michalsidzej) [reply packages/discovery/src/discovery/engine/DiscoveryEngine.ts:10]: 9246c51a

Comment (michalsidzej) [reply packages/discovery/src/discovery/engine/DiscoveryEngine.ts:10]: 9246c51a

File: packages/backend/src/config/chains.ts
Comment (michalsidzej) [reply packages/backend/src/config/chains.ts]: b9849c39

**Review by antooni — APPROVED** [review]

File: packages/backend/src/config/chains.ts
Comment (sdlyy) [inline packages/backend/src/config/chains.ts:2]:
```suggestion
import { assert, ChainId, UnixTime } from '@l2beat/shared-pure'
```
Prolly miss-import

**Review by sdlyy — APPROVED** [review]
Looks solid - gj, please fix miss-import before merging 🙏🏻

**Pull request merged**
